File: scala/day1/part1.scala
import scala.io.Source
object Part1 extends App {
    def getInputs(filepath: String): List[Int] =    
        Source.fromFile(filepath).getLines().map(_.toInt).toList

    def find2NumbersThatSumTo(goal: Int, numbers: List[Int]): Option[(Int, Int)] = {
        def checkAt(start: Int, end: Int): Option[(Int, Int)] = {
            if (start == end) {
                return None
            }

            val sum = numbers(start) + numbers(end)
            if (sum == goal) {
                return Some((numbers(start), numbers(end)))
            } else if (sum < goal) {
                return checkAt(start + 1, end)
            } else {
                return checkAt(start, end - 1)
            }
        }

        checkAt(0, numbers.length - 1)
    }

    val file = "../../inputs/day1.txt"
    val goal = 2020
    val inputs = getInputs(file).sorted

    (find2NumbersThatSumTo(goal, inputs)) match {
        case (Some(x)) => println(f"${x._1}, ${x._2} -> ${x._1 * x._2}") 
        case (None) => println("No solution")
    }
}

File: python/day1/part1.py
import functools

def getInput(filename):
    f = open(filename)
    numbers = f.readlines()
    f.close()
    return list(map(lambda x: int(x), numbers))

def find2NumbersThatSumTo(goal, numbers, start):
    end = len(numbers) - 1
    sum = numbers[start] + numbers[end]
    while sum != goal and end > start:
        if sum < goal:
            start += 1
        else:
            end -= 1
        sum = numbers[start] + numbers[end]
    if sum == goal:
        return (numbers[start], numbers[end])
    else:
        return False

def main():
    filename = "../../inputs/day1.txt"
    numbers = sorted(getInput(filename))
    result = find2NumbersThatSumTo(2020, numbers, 0)
    if (result == False):
        print("No result")
    else:
        print(*result)
        print(functools.reduce(lambda a,b: a*b, result))

main()

